### ISSUE TITLE
Adding `mapOptions` has observable creates problem in callback hooks

### ISSUE THREAD

**Issue by SamvelRaja** [SamvelRaja/google-maps-addon#27]
Since the `mapOptions` is an observable property, when we use the addon in an component, this forces us to create the `mapOptions` object has mutable object to handle the CB function, therefore, any changes happened in the component property itself creates the `mapOptions` to update.

say,
converting the mapMessage from [mousemove](https://github.com/SamvelRaja/SamvelRaja.github.io/blob/master/site/app/demos/googlemaps/map-component/component.js#L23) to

[mouseout](https://github.com/SamvelRaja/SamvelRaja.github.io/blob/master/site/app/demos/googlemaps/map-component/component.js#L26) triggers back the the component to re-render

[DEMO](http://samvelraja.github.io/#/demos/googlemaps)

Must find a fix ASAP.

One thought is to move-out the CB properties out of mapOptions and providing them as seperate options.

Goal, should not revert back the **observer** since it is a good enhancement.

@erkie  your thoughts are welcome.

**Comment by erkie** [comment]
I checked out your repo and found that the problem seems to be that you are modifying the `inputObj` property on every render. So every time you change `mapMessage` it triggers `willRender` which causes `mapOptions` to change and that to re-render the map, which causes the flickering.

If you change the `inputObj` to be a computed property it works smoothly, which I think is an OK solution for that specific problem. [Example of `component.js` that does not flicker](https://gist.github.com/erkie/f50c03460da541662be8)

However, the complete re-rendering of the map does cause the bounce animation to appear and timeout to trigger once all over again, causing another type of "flicker".

I was looking into implementing some sort of smart diffing for the markers array. Maybe that's the way forward? My reasoning for the observer on `mapOptions` was change the markers array. If we could easily diff it and animate in new ones, remove old ones, and do nothing to unchanged ones, the re-animation problem would not occur. 

The problem I ran into was that when trying to find old and new markers I compared references. The references change everytime since usually the markers array is recreated on every render of `mapOptions`, so diffing is impossible. How about we do a React-style `key`-attribute on the `markers`-array? So we can easily diff them. So for example:

``` js
mapOptions: Ember.computed("places", function() {
     return {
          markers: this.get("places").map((place, index) => {
                return {
                     key: place.get("id"),
                     latitude: place.get("latitude"),
                     longitude: place.get("longitude"),
                     timeout: index * 1000,
                     animation: "BOUNCE"
                }
          })
     }
});
```

This would be nice so we also can add some more options to the markers array, like how to transition a marker between lats/longs, `onRemove/onAdd/onChange`-callbacks per marker, etc.

If you think that's OK I can try to implement something like that today?
